src/nvim/os/shell.c

```diff
@@ -93,31 +93,37 @@ void shell_free_argv(char **argv)
 }
 
 /// Joins shell arguments from `argv` into a new string.
+/// If the result is too long it is truncated with ellipsis ("...").
 ///
 /// @returns[allocated] `argv` joined to a string.
 char *shell_argv_to_str(char **const argv)
   FUNC_ATTR_NONNULL_ALL
 {
   size_t n = 0;
   char **p = argv;
-  if (p == NULL) {
-    return 0;
+  char *rv = xcalloc(256, sizeof(*rv));
+  const size_t maxsize = 256 * sizeof(*rv);
+  if (*p == NULL) {
+    rv[0] = '\0';
   }
-
-  const size_t maxsize = 256;
-  char *rv = xcalloc(maxsize, sizeof(*rv));
-  while (*p != NULL && n < maxsize) {
-    n += xstrlcat(rv, "'", maxsize);
-    n += xstrlcat(rv, *p, maxsize);
-    n += xstrlcat(rv, (*(p + 1) == NULL) ? "'" : "' ", maxsize);
+  while (*p != NULL) {
+    if ((*p)[0] != '\0') {  // Skip empty strings.
+      xstrlcat(rv, "'", maxsize);
+      xstrlcat(rv, *p, maxsize);
+      n = xstrlcat(rv,  "' ", maxsize);
+      if (n >= maxsize) {
+        break;
+      }
+    }
     p++;
   }
-  if (n >= maxsize) {
+  if (n < maxsize) {
```

Comment:
`n <= maxsize`:
If `n == maxsize` then the last space that should be removed was already replaced by `xstrlcat()` with `'\0'`. This case fits better here than in the other branch.